### ISSUE TITLE
Can't generate verbatim manifest without filters in AnVIL

### ISSUE THREAD

**Issue by achave11-ucsc** [DataBiosphere/azul#6108]
https://github.com/DataBiosphere/azul/issues/6108#issuecomment-2035944695

**Comment by dsotirho-ucsc** [comment]
Assignee to consider next steps.

**Comment by hannes-ucsc** [comment]
The generation simply times out.

Our standard approach to this problem is to partition the manifest as we do for the compact manifest, but the verbatim manifest formats complicate this by having to ensure that each replica is only written once. The AvroPFB verbatim manifest further complicates this with the forward-only reference constraint that Terra imposes. We are currently working around that constraint by not exposing relations/links in the AvroPFB schema we generate.

The naive approach to ensuring uniqueness of replicas in the generated manifest is to use a set of already emitted replicas. Since each partition in a partitioned manifest runs in a different Lambda invocation and potentially a different execution context, we would have to persist that set between invocations.  My original design already describes [optimizations](https://github.com/DataBiosphere/azul/issues/2693#user-content-4943) to reduce the size of the set (not tracking replicas with just one hub, tracking hubs instead of replicas). A smaller set is obviously faster to read and write.

I don't have a good solution at this time. Interestingly, the verbatim JSONL manifest for HCA does not timeout (without links, we don't know if adding links breaks that). I don't think there is much of a use case for a all-inclusive, unfiltered AvroPFB manifest. The purpose of that manifest is exporting it to Terra. The resulting Terra workspace would be huge for both AnVIL and HCA.

**Comment by dsotirho-ucsc** [comment]
Moving ticket to `Triage` to consolidate with tickets similar to this.

**Comment by dsotirho-ucsc** [comment]
Also note #3892 and #6123

**Comment by hannes-ucsc** [comment]
#3892 and #6123 refer to the content-hash computation timing out, so it never gets to even start the generation. In fact, the #3892 was closed as a duplicate of #6123 for which we have potential solution/workaround planned.

**Comment by nadove-ucsc** [comment]
This was solved for the verbatim  JSONL via https://github.com/DataBiosphere/azul/issues/7190. It it still valid for the verbatim PFB manifest.

**Comment by hannes-ucsc** [comment]
Added blocker #6123 and put it back to Triage. Once blocker lands, we'll revisit this. The fact that we solved this for JSONL (which is more generally useful than the niche PFB) is a big win.

**Comment by dsotirho-ucsc** [comment]
(partial) PR https://github.com/DataBiosphere/azul/pull/7258 for blocking issue https://github.com/DataBiosphere/azul/issues/6123 has been merged.

**Comment by dsotirho-ucsc** [comment]
Spike to attempt to reproduce in anvilprod once the blocker (#6123) lands there.

**Comment by achave11-ucsc** [comment]
@hannes-ucsc: "Reschedule spike as soon as bundle notification are disabled in `anvilprod` which will happen as a partial PR against blocker #6123. Before we can create that partial PR, we need to finish the [demo](https://github.com/DataBiosphere/azul/issues/6123#issuecomment-3974826735) for the previous partial PR #7258."

**Comment by dsotirho-ucsc** [comment]
@hannes-ucsc: "Another partial PR (#7837) against issue #6123 has landed. There is [another demo to be done](https://github.com/DataBiosphere/azul/issues/6123#issuecomment-4025487708) in production before we can finally schedule the spike."

**Comment by dsotirho-ucsc** [comment]
We were planning to promote PR #7837, but didn't end up doing so. So it will be promoted next week.

**Comment by achave11-ucsc** [comment]
Spike to attempt to reproduce in anvilprod now that the blocker ([#6123](https://github.com/DataBiosphere/azul/issues/6123)) has landed.

**Comment by dsotirho-ucsc** [comment]
Edit: Note that the reproduction below is for a JSONL manifest, not PFB.

----

> Spike to attempt to reproduce in anvilprod now that the blocker ([#6123](https://github.com/DataBiosphere/azul/issues/6123)) has landed.

Unable to reproduce failure. A request with no filters for a verbatim JSONL manifest completed just after one hour.

```
$ ll -h anvil-manifest-3ef67165-d384-519e-be40-972b72ee9e06.4f633b65-6b62-5c24-8659-2853fbb10c51.jsonl
-rw-r--r--@ 1 daniel  staff   6.5G Mar 26 12:07 anvil-manifest-3ef67165-d384-519e-be40-972b72ee9e06.4f633b65-6b62-5c24-8659-2853fbb10c51.jsonl

$ wc -l anvil-manifest-3ef67165-d384-519e-be40-972b72ee9e06.4f633b65-6b62-5c24-8659-2853fbb10c51.jsonl
 10271309 anvil-manifest-3ef67165-d384-519e-be40-972b72ee9e06.4f633b65-6b62-5c24-8659-2853fbb10c51.jsonl
```

## /aws/lambda/azul-service-anvilprod    

| @timestamp | @message |
| --- | --- |
| 2026-03-26 17:34:46.307 | START RequestId: e4110b33-08a8-44cc-babd-64dd64c5c254 Version: 24 |
| 2026-03-26 17:34:46.308 | [INFO] 2026-03-26T17:34:46.308Z e4110b33-08a8-44cc-babd-64dd64c5c254 azul.chalice Received PUT request for &#x27;&#x2f;fetch&#x2f;manifest&#x2f;files&#x27;, with {&quot;query&quot;: {&quot;catalog&quot;: &quot;anvil13&quot;, &quot;format&quot;: &quot;verbatim.jsonl&quot;}, &quot;headers&quot;: {&quot;accept&quot;: &quot;application&#x2f;json&quot;, &quot;accept-encoding&quot;: &quot;gzip, deflate, br, zstd&quot;, &quot;accept-language&quot;: &quot;en-US,en;q=0.9&quot;, &quot;authorization&quot;: &quot;Bearer ya29.a0Aa7&lt;REDACTED&gt;&quot;, &quot;cloudfront-forwarded-proto&quot;: &quot;https&quot;, &quot;cloudfront-is-desktop-viewer&quot;: &quot;true&quot;, &quot;cloudfront-is-mobile-viewer&quot;: &quot;false&quot;, &quot;cloudfront-is-smarttv-viewer&quot;: &quot;false&quot;, &quot;cloudfront-is-tablet-viewer&quot;: &quot;false&quot;, &quot;cloudfront-viewer-asn&quot;: &quot;7922&quot;, &quot;cloudfront-viewer-country&quot;: &quot;US&quot;, &quot;host&quot;: &quot;service.explore.anvilproject.org&quot;, &quot;origin&quot;: &quot;https:&#x2f;&#x2f;service.explore.anvilproject.org&quot;, &quot;priority&quot;: &quot;u=1, i&quot;, &quot;referer&quot;: &quot;https:&#x2f;&#x2f;service.explore.anvilproject.org&#x2f;swagger&#x2f;index.html&quot;, &quot;sec-ch-ua&quot;: &quot;\&quot;Chromium\&quot;;v=\&quot;146\&quot;, \&quot;Not-A.Brand\&quot;;v=\&quot;24\&quot;, \&quot;Google Chrome\&quot;;v=\&quot;146\&quot;&quot;, &quot;sec-ch-ua-mobile&quot;: &quot;?0&quot;, &quot;sec-ch-ua-platform&quot;: &quot;\&quot;macOS\&quot;&quot;, &quot;sec-fetch-dest&quot;: &quot;empty&quot;, &quot;sec-fetch-mode&quot;: &quot;cors&quot;, &quot;sec-fetch-site&quot;: &quot;same-origin&quot;, &quot;sec-gpc&quot;: &quot;1&quot;, &quot;user-agent&quot;: &quot;Mozilla&#x2f;5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit&#x2f;537.36 (KHTML, like Gecko) Chrome&#x2f;146.0.0.0 Safari&#x2f;537.36&quot;, &quot;via&quot;: &quot;2.0 6c393e4d7ff07e3c2467f7a0cb584a8e.cloudfront.net (CloudFront)&quot;, &quot;x-amz-cf-id&quot;: &quot;-IGyShmyD6Ie2345wEMXMWIU51ddXIisGL9w3491SfB2jC9P5XMU6g==&quot;, &quot;x-amzn-trace-id&quot;: &quot;Root=1-69c56e36-5831e73c5af5b1ab1537e3f1&quot;, &quot;x-forwarded-for&quot;: &quot;73.189.51.68, 15.158.5.138&quot;, &quot;x-forwarded-port&quot;: &quot;443&quot;, &quot;x-forwarded-proto&quot;: &quot;https&quot;}}. |
| 2026-03-26 17:34:46.308 | [INFO] 2026-03-26T17:34:46.308Z e4110b33-08a8-44cc-babd-64dd64c5c254 azul.chalice … without a request body |
| | ... |
| 2026-03-26 17:34:48.107 | [INFO] 2026-03-26T17:34:48.107Z e4110b33-08a8-44cc-babd-64dd64c5c254 azul.service.manifest_service Cached manifest not found: ManifestKey(catalog=&#x27;anvil13&#x27;, format=&lt;ManifestFormat.verbatim_jsonl: &#x27;verbatim.jsonl&#x27;&gt;, manifest_hash=UUID(&#x27;3ef67165-d384-519e-be40-972b72ee9e06&#x27;), source_hash=UUID(&#x27;4f633b65-6b62-5c24-8659-2853fbb10c51&#x27;)) |



## /aws/lambda/azul-service-anvilprod-manifest    

| @timestamp | @message |
| --- | --- |
| 2026-03-26 17:34:52.944 | [INFO] 2026-03-26T17:34:52.944Z 40302d63-2e2c-4902-bd29-119645b15c7a azul.service.manifest_service Cached manifest not found: ManifestKey(catalog=&#x27;anvil13&#x27;, format=&lt;ManifestFormat.verbatim_jsonl: &#x27;verbatim.jsonl&#x27;&gt;, manifest_hash=UUID(&#x27;3ef67165-d384-519e-be40-972b72ee9e06&#x27;), source_hash=UUID(&#x27;4f633b65-6b62-5c24-8659-2853fbb10c51&#x27;)) |
| | ... |
| 2026-03-26 18:36:15.361 | [INFO] 2026-03-26T18:36:15.361Z 008420a5-0000-406b-bee2-4efb61e29c9e azul.boto3 s3.UploadPart: Making PUT request to https:&#x2f;&#x2f;s3.amazonaws.com&#x2f;edu-ucsc-gi-platform-anvil-prod-storage-anvilprod.us-east-1&#x2f;manifests&#x2f;3ef67165-d384-519e-be40-972b72ee9e06.4f633b65-6b62-5c24-8659-2853fbb10c51.jsonl?uploadId=gY1zmw&lt;REDACTED&gt;&amp;partNumber=66 |
| 2026-03-26 18:36:15.361 | [INFO] 2026-03-26T18:36:15.361Z 008420a5-0000-406b-bee2-4efb61e29c9e azul.boto3 s3.UploadPart: … with a request body of type (&lt;class &#x27;botocore.httpchecksum.AwsChunkedWrapper&#x27;&gt;) |
| 2026-03-26 18:36:15.712 | [INFO] 2026-03-26T18:36:15.712Z 008420a5-0000-406b-bee2-4efb61e29c9e azul.boto3 s3.UploadPart: Got 200 response |
| 2026-03-26 18:36:15.712 | [INFO] 2026-03-26T18:36:15.712Z 008420a5-0000-406b-bee2-4efb61e29c9e azul.boto3 s3.UploadPart: … without a response body |
| 2026-03-26 18:36:15.716 | [INFO] 2026-03-26T18:36:15.716Z 008420a5-0000-406b-bee2-4efb61e29c9e azul.boto3 s3.CompleteMultipartUpload: Making POST request to https:&#x2f;&#x2f;s3.amazonaws.com&#x2f;edu-ucsc-gi-platform-anvil-prod-storage-anvilprod.us-east-1&#x2f;manifests&#x2f;3ef67165-d384-519e-be40-972b72ee9e06.4f633b65-6b62-5c24-8659-2853fbb10c51.jsonl?uploadId=gY1zmw&lt;REDACTED&gt; |
| 2026-03-26 18:36:15.716 | [INFO] 2026-03-26T18:36:15.716Z 008420a5-0000-406b-bee2-4efb61e29c9e azul.boto3 s3.CompleteMultipartUpload: … with a request body of length 5832 starting in b&#x27;&lt;CompleteMultipartUpload xmlns=&quot;http:&#x2f;&#x2f;s3.amazonaws.com&#x2f;doc&#x2f;2006-03-01&#x2f;&quot;&gt;&lt;Part&gt;&lt;PartNumber&gt;1&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;2cccd749eba855dc5400e80b4d10fd65&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;2&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;da09c346325488b58ea14bea3d7fd91f&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;3&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;be6178c0794c42e00084e40ac3ff350c&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;4&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;26ae91a2f7dcc1efff90289a9f01ed64&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;5&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;102db6f195686161ac2f485874fcb337&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;6&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;9e2c120825e0d22c8277f71bf969c035&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;7&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;851ca0f425022a4c40a4fcecbf9f0b65&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;8&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;4ea01247dd8263891a7ebae1a84d89ef&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;9&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;fb809928aad38b85dc111ee7f2db8f56&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;10&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;d63bb30efccbc078f85784e543351250&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;&lt;Part&gt;&lt;PartNumber&gt;11&lt;&#x2f;PartNumber&gt;&lt;ETag&gt;&quot;741d2c7ddaa4bf67b1afc32f6b6b2b03&quot;&lt;&#x2f;ETag&gt;&lt;&#x2f;Part&gt;...&#x27; |
| 2026-03-26 18:36:15.889 | [INFO] 2026-03-26T18:36:15.889Z 008420a5-0000-406b-bee2-4efb61e29c9e azul.boto3 s3.CompleteMultipartUpload: Got 200 response |
| 2026-03-26 18:36:15.889 | [INFO] 2026-03-26T18:36:15.889Z 008420a5-0000-406b-bee2-4efb61e29c9e azul.boto3 s3.CompleteMultipartUpload: … with a response body of length 665 being b&#x27;&lt;?xml version=&quot;1.0&quot; encoding=&quot;UTF-8&quot;?&gt;\n\n&lt;CompleteMultipartUploadResult xmlns=&quot;http:&#x2f;&#x2f;s3.amazonaws.com&#x2f;doc&#x2f;2006-03-01&#x2f;&quot;&gt;&lt;Location&gt;https:&#x2f;&#x2f;s3.amazonaws.com&#x2f;edu-ucsc-gi-platform-anvil-prod-storage-anvilprod.us-east-1&#x2f;manifests%2F3ef67165-d384-519e-be40-972b72ee9e06.4f633b65-6b62-5c24-8659-2853fbb10c51.jsonl&lt;&#x2f;Location&gt;&lt;Bucket&gt;edu-ucsc-gi-platform-anvil-prod-storage-anvilprod.us-east-1&lt;&#x2f;Bucket&gt;&lt;Key&gt;manifests&#x2f;3ef67165-d384-519e-be40-972b72ee9e06.4f633b65-6b62-5c24-8659-2853fbb10c51.jsonl&lt;&#x2f;Key&gt;&lt;ETag&gt;&quot;d6c3e594517b45004bbc63db2cc0cdac-66&quot;&lt;&#x2f;ETag&gt;&lt;ChecksumCRC64NVME&gt;qBVnoi8FdaQ=&lt;&#x2f;ChecksumCRC64NVME&gt;&lt;ChecksumType&gt;FULL_OBJECT&lt;&#x2f;ChecksumType&gt;&lt;&#x2f;CompleteMultipartUploadResult&gt;&#x27; |

**Comment by achave11-ucsc** [comment]
@hannes-ucsc to review spike results.

**Comment by hannes-ucsc** [comment]
This has been fixed for JSONL which the vastly more useful format, anyways. PFB is a niche format, and a cumbersome one at that. The only reason we implemented it was to facilitate exports to Terra and Terra adds even more complicated constraints to this already difficult to implement format. 

https://github.com/DataBiosphere/azul/issues/6108#issuecomment-2035944695

> I don't have a good solution at this time. Interestingly, the verbatim JSONL manifest for HCA does not timeout (without links, we don't know if adding links breaks that). I don't think there is much of a use case for a all-inclusive, unfiltered AvroPFB manifest. The purpose of that manifest is exporting it to Terra. The resulting Terra workspace would be huge for both AnVIL and HCA.

Yes, no feasible plan for resolution exists at this time. The BI folks have only have ever tested the verbatim PFB export for individual datasets at the most. We have not heard complaints about this from anyone else, either. Iceboxing for now, given the resolution
